Make sure to wait before restarting the remote builder

billyb2 opened the pull request:
### Change Summary

What and Why:

If the remote builder is in the 'created' or 'migrating' state, then we won't be able to restart it. By making sure to wait for one of those two states, we can avoid errors like:

`failed to fetch an image or build from source: failed to restart VM asfijsdoif: failed_precondition: unable to restart machine, not currently started or stopped`

How:
make sure that the builder machine is either in the started or stopped state before calling the 'restart' API

Related to:

---

### Documentation

- [x] Fresh Produce (i'm gonna fix a few more remote builder issues before FP)
- [ ] In superfly/docs, or asked for help from docs team
- [ ] n/a

billyb2 committed:
Make sure to wait before restarting the remote builder

If the remote builder is in the 'created' or 'migrating' state, then we
won't be able to restart it. By making sure to wait for one of those two
states, we can avoid errors like:

`failed to fetch an image or build from source: failed to restart VM asfijsdoif: failed_precondition: unable to restart machine, not currently started or stopped`

jipperinbham commented on internal/build/imgsrc/ensure_builder_test.go:
```diff
@@ -275,16 +276,24 @@ func TestRestartBuilderMachine(t *testing.T) {
 		},
 		WaitFunc: func(ctx context.Context, machine *fly.Machine, state string, timeout time.Duration) (err error) {
+			if state == "started" || state == "stopped" {
+				waitedForStartOrStop = true
```
I don't understand how this test is asserting the logic correctly since the initial state of `waitedForStartOrStop` is `true`.

billyb2 replied:
oh, yea i should set that to false by default 🤦🏾

billyb2 committed:
fixed a test

billyb2 committed:
simplified WaitForAnyMachineState

channels are better than that disgusting mutex garbage i was doing

jipperinbham commented on internal/build/imgsrc/ensure_builder.go:
```diff
@@ -409,6 +410,8 @@ func restartBuilderMachine(ctx context.Context, builderMachine *fly.Machine) err
 
 	flapsClient := flapsutil.ClientFromContext(ctx)
 
+	mach.WaitForAnyMachineState(ctx, builderMachine, []string{"started", "stopped"}, 60*time.Second, nil)
```
Should we be checking the return values? Or are we restarting the machine even if it's already in `started` state which seems to be the desired state?

billyb2 replied:
yes we absolutely should. thank you for pointing this out :)

billyb2 committed:
start waitedForStartOrStop to false

silly testing bug

billyb2 committed:
Change restartBuilder to return cleanly when already started

Basically, if the builder is in the 'created' state for example, and we
transition into 'started', there's no reason to start it.

billyb2 committed:
don't return err = nil if we can't find the correct state

that's dumb. instead we should just return a standard WaitTimeoutErr if
we can't reach any state.

billyb2 committed:
retry starting up the remote builder machine

billyb2 committed:
cleanup code a bit

switch statements are the future

Co-authored-by: JP 829493+jipperinbham@users.noreply.github.com

billyb2 committed:
fix tests

billyb2 commented on internal/command/deploy/plan.go:
```diff
@@ -493,11 +494,16 @@ func (md *machineDeployment) updateMachineWChecks(ctx context.Context, oldMachin
 		err = lm.WaitForHealthchecksToPass(ctx, md.waitTimeout)
-		if err != nil {
+		switch {
+		case errors.Is(err, context.DeadlineExceeded):
```
didn't mean to commit this, but this is a small fix to retry health checks correctly.
